src/smt/term_formula_removal.cpp

```diff
@@ -91,91 +87,133 @@ Node RemoveTermFormulas::run(TNode node, std::vector<Node>& output,
   int cv = cacheVal( inQuant, inTerm );
   std::pair<Node, int> cacheKey(node, cv);
   NodeManager *nodeManager = NodeManager::currentNM();
-  ITECache::const_iterator i = d_iteCache.find(cacheKey);
-  if(i != d_iteCache.end()) {
+  TermFormulaCache::const_iterator i = d_tfCache.find(cacheKey);
+  if (i != d_tfCache.end())
+  {
     Node cached = (*i).second;
     Debug("ite") << "removeITEs: in-cache: " << cached << endl;
     return cached.isNull() ? Node(node) : cached;
   }
 
 
-  // If an ITE, replace it
   TypeNode nodeType = node.getType();
   Node skolem;
   Node newAssertion;
   if(node.getKind() == kind::ITE) {
-    if(!nodeType.isBoolean() && (!inQuant || !node.hasBoundVar())) {
-      // Make the skolem to represent the ITE
-      skolem = nodeManager->mkSkolem("termITE", nodeType, "a variable introduced due to term-level ITE removal");
-
-      // The new assertion
-      newAssertion =
-        nodeManager->mkNode(kind::ITE, node[0], skolem.eqNode(node[1]),
-                            skolem.eqNode(node[2]));
+    // If an ITE, replace it
+    if (!nodeType.isBoolean() && (!inQuant || !node.hasBoundVar()))
+    {
+      skolem = getSkolemForNode(node);
+      if (skolem.isNull())
+      {
+        // Make the skolem to represent the ITE
+        skolem = nodeManager->mkSkolem(
+            "termITE",
+            nodeType,
+            "a variable introduced due to term-level ITE removal");
+        d_skolem_cache.insert(node, skolem);
+
+        // The new assertion
+        newAssertion = nodeManager->mkNode(
+            kind::ITE, node[0], skolem.eqNode(node[1]), skolem.eqNode(node[2]));
+      }
     }
   }
-  //if a lambda, do lambda-lifting
-  if( node.getKind() == kind::LAMBDA && !inQuant ){
-    // Make the skolem to represent the lambda
-    skolem = nodeManager->mkSkolem("lambdaF", nodeType, "a function introduced due to term-level lambda removal");
-
-    // The new assertion
-    std::vector< Node > children;
-    // bound variable list
-    children.push_back( node[0] );
-    // body
-    std::vector< Node > skolem_app_c;
-    skolem_app_c.push_back( skolem );
-    for( unsigned i=0; i<node[0].getNumChildren(); i++ ){
-      skolem_app_c.push_back( node[0][i] );
+  else if (node.getKind() == kind::LAMBDA)
+  {
+    // if a lambda, do lambda-lifting
+    if (!inQuant)
+    {
+      skolem = getSkolemForNode(node);
+      if (skolem.isNull())
+      {
+        // Make the skolem to represent the lambda
+        skolem = nodeManager->mkSkolem(
+            "lambdaF",
+            nodeType,
+            "a function introduced due to term-level lambda removal");
+        d_skolem_cache.insert(node, skolem);
+
+        // The new assertion
+        std::vector<Node> children;
+        // bound variable list
+        children.push_back(node[0]);
+        // body
+        std::vector<Node> skolem_app_c;
+        skolem_app_c.push_back(skolem);
+        for (unsigned i = 0; i < node[0].getNumChildren(); i++)
```

Comment:
Cache `getNumChildren()` and use a node builder.

Reply:
Or use `skolem_app_c.insert(skolem_app_c.end(), node[0].begin(), node[0].end());`